mempool/src/validator.rs

```diff
@@ -0,0 +1,325 @@
+            return Err(PreFlightError::GasLimitReached);
+        }
+
+        if gas_limit > u64::MAX as u128 {
+            return Err(PreFlightError::GasU64Overflow);
+        }
+
+        let gas_limit_u256 = u256::new(gas_limit);
+        let gas_price = u256::new(tx_gas_price.unwrap_or(0));
+        gas_limit_u256
+            .checked_mul(gas_price)
+            .ok_or(PreFlightError::MaxFeePerGasU256Overflow)?;
+
+        Ok(())
+    }
+
+    async fn validate_sender_balance(
+        tx: &ExecuteRequest,
+        tx_input_len: usize,
+        tx_gas_limit: u128,
+        tx_gas_price: Option<u128>,
+        neon_api: &NeonApi,
+    ) -> Result<(), PreFlightError> {
+        let sender_address = tx.recover_signer()?;
+        let balance_address = BalanceAddress {
+            address: sender_address.to_neon(),
+            chain_id: tx.fallback_chain_id(),
+        };
+        let sender_balance = neon_api.get_balance(balance_address, None).await?; // todo: which commitment to use?
+        let required_balance = U256::from(tx_gas_price.unwrap_or(0)) * U256::from(tx_gas_limit)
+            + to_u256(tx_value(tx)?);
+
+        if required_balance <= sender_balance {
+            return Ok(());
+        }
+
+        let message = if tx_input_len == 0 {
+            "insufficient funds for transfer"
+        } else {
+            "insufficient funds for gas * price + value"
+        };
+
+        Err(PreFlightError::InsufficientFundsForTransfer(
+            message,
+            sender_address,
+            sender_balance,
+            required_balance,
+        ))
+    }
+
+    fn validate_tx_gas_price(
+        chain_id: Option<u64>,
+        tx_gas_price: Option<u128>,
+        price: &GasPriceModel,
+    ) -> Result<(), PreFlightError> {
+        let min_gas_price = price.min_acceptable_gas_price;
+        if let Some(gas_price) = tx_gas_price {
+            let gas_price = U128::from(gas_price);
+            if gas_price < min_gas_price && chain_id.is_some() {
+                return Err(PreFlightError::Underpriced(gas_price, min_gas_price));
+            }
+        }
+
+        Ok(())
+    }
+
+    fn validate_underpriced_tx_wo_chain_id(
+        chain_id: Option<u64>,
+        tx_gas_price: Option<u128>,
+        price: &GasPriceModel,
+    ) -> Result<(), PreFlightError> {
+        if chain_id.is_some() {
+            return Ok(());
+        }
+        if let Some(gas_price) = tx_gas_price {
+            let gas_price = U128::from(gas_price);
+            if gas_price < price.min_wo_chain_id_acceptable_gas_price {
+                return Err(PreFlightError::UnderpricedWoChainId);
+            }
+        }
+
+        Ok(())
+    }
+}
+
+fn to_u256(v: alloy_primitives::U256) -> U256 {
+    let limbs = v.as_limbs();
+    let limb0 = (limbs[1] as u128) << 64 | (limbs[0] as u128); // Lower u128
+    let limb1 = (limbs[3] as u128) << 64 | (limbs[2] as u128); // Upper u128
+
+    U256([limb0, limb1])
+}
```

Comment:
This is already implemented in common
https://github.com/bestarch-ae/neon-proxy/blob/a02611055d2620aaec00b6992f31d39f55ac0159/common/src/convert.rs#L64-L70